api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -165,35 +169,57 @@ public void close() {
     }
   }
 
-  private void scheduleTokensRenewal(Tokens currentTokens) {
-    Instant now = Instant.now();
+  private void maybeScheduleTokensRenewal(Tokens currentTokens) {
+    Instant now = clock.get();
+    if (idle(now)) {
+      sleeping.set(true);
+      LOGGER.debug("Sleeping...");
+    } else {
+      scheduleOrExecuteTokensRenewal(currentTokens, now, MIN_REFRESH_DELAY);
+    }
+  }
+
+  private void scheduleOrExecuteTokensRenewal(
+      Tokens currentTokens, Instant now, Duration minRefreshDelay) {
     Instant accessExpirationTime =
         tokenExpirationTime(now, currentTokens.getAccessToken(), defaultAccessTokenLifespan);
     Instant refreshExpirationTime =
         tokenExpirationTime(now, currentTokens.getRefreshToken(), defaultRefreshTokenLifespan);
     Duration delay =
-        nextDelay(now, accessExpirationTime, refreshExpirationTime, refreshSafetyWindow);
-    LOGGER.debug("Scheduling token refresh in {}", delay);
-    tokenRefreshFuture =
-        executor.schedule(
-            () -> {
-              currentTokensStage = renewTokens();
-              currentTokensStage.thenAccept(this::scheduleTokensRenewal);
-            },
-            delay.toMillis(),
-            TimeUnit.MILLISECONDS);
+        nextDelay(
+            now, accessExpirationTime, refreshExpirationTime, refreshSafetyWindow, minRefreshDelay);
+    if (delay.compareTo(MIN_REFRESH_DELAY) < 0) {
+      LOGGER.debug("Refreshing tokens immediately");
+      renewTokens();
+    } else {
+      LOGGER.debug("Scheduling token refresh in {}", delay);
+      tokenRefreshFuture =
+          executor.schedule(this::renewTokens, delay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void renewTokens() {
+    CompletionStage<Tokens> oldTokensStage = currentTokensStage;
```

Comment:
Nit: Should `oldTokansStage` be inlined?

Reply:
IJ gives a warning: "Non-atomic operation on volatile field 'currentTokensStage'".

Reply:
Oh, interesting.